File: 2 - Work/Phase 3 Economic Merge and MICE Imputation/Phase_3.py
# Purpose: Complete Phase 3 pipeline - MICE imputation (m=100, LightGBM Random
#          Forest) on the Phase 2 acreage-matched dataset to fill missing
#          osm_acreage and Baseline_Value_Per_Acre values.
# Inputs:  Phase 2 Spatial Polygons and True Acreage/Data/python/
#            Py_Phase2_Acreage_Matched.csv
# Outputs: Data/python/Py_Imputed_Dataset_{1..100}.csv
#          Data/python/Py_Rubins_Rules_Summary.csv
#          Data/python/Py_National_Acreage_Summary.csv


# === 1. LIBRARIES ===

import gc
import multiprocessing
import pathlib

import miceforest as mf
import numpy as np
import pandas as pd

import warnings
warnings.filterwarnings("ignore", category=pd.errors.PerformanceWarning)


# === 2. GLOBALS & PATHS ===

SCRIPT_DIR = pathlib.Path(__file__).parent
INPUT_CSV  = (
    SCRIPT_DIR.parent
    / "Phase 2 Spatial Polygons and True Acreage"
    / "Data" / "python"
    / "Py_Phase2_Acreage_Matched.csv"
)
OUT_DIR         = SCRIPT_DIR / "Data" / "python"
OUT_RUBINS_CSV  = OUT_DIR / "Py_Rubins_Rules_Summary.csv"
OUT_ACREAGE_CSV = OUT_DIR / "Py_National_Acreage_Summary.csv"

M              = 100
IMPUTE_COLS    = ["osm_acreage", "Baseline_Value_Per_Acre"]
PREDICTOR_COLS = ["Holes", "Ownership_Type", "county_type", "Longitude", "Latitude"]

# Reserve one core for the OS; caps at 23 on a 24-thread 3900XT
N_CORES = max(1, multiprocessing.cpu_count() - 1)


# === 3. FUNCTIONS ===

def run_imputation(input_csv, out_dir, m_datasets=100):
    # 1. Load ----------------------------------------------------------------
    print("--- 1  Loading Phase 2 acreage-matched dataset ---")

    if not input_csv.exists():
        raise FileNotFoundError(f"Input file not found: {input_csv}")

    out_dir.mkdir(parents=True, exist_ok=True)

    acreage_df = pd.read_csv(input_csv)
    print(f"    Rows: {len(acreage_df):,}")
    print(
        f"    Missing osm_acreage:            "
        f"{acreage_df['osm_acreage'].isna().sum():,}"
    )
    print(
        f"    Missing Baseline_Value_Per_Acre: "
        f"{acreage_df['Baseline_Value_Per_Acre'].isna().sum():,}"
    )

    # 2. Prepare -------------------------------------------------------------
    print("--- 2  Preparing imputation frame ---")

    acreage_df["Holes"]        = pd.to_numeric(acreage_df["Holes"], errors="coerce")
    acreage_df["Ownership_Type"] = acreage_df["Ownership_Type"].astype("category")
    acreage_df["county_type"]  = acreage_df["county_type"].astype("category")

    model_cols = PREDICTOR_COLS + IMPUTE_COLS
    imp_df = acreage_df[model_cols].copy()

    # 3. Run miceforest -------------------------------------------------------
    print(
        f"--- 3  Running miceforest "
        f"(m={m_datasets}, LightGBM backend, n_jobs={N_CORES}) ---"
    )
    # [METHODOLOGY] miceforest MICE - LightGBM Random Forest, m=100 datasets,
    #               random_state=42 for reproducibility (Van Buuren 2018)
    imputed_list = mf.ImputationKernel(
        data=imp_df,
        num_datasets=m_datasets,
        random_state=42,
    )
    # [METHODOLOGY] 10 iterations per dataset; n_jobs parallelises LightGBM trees
    imputed_list.mice(iterations=10, verbose=True, n_jobs=N_CORES)

    # 4. Extract & save each imputed dataset ---------------------------------
    print(f"\n--- 4  Saving {m_datasets} imputed datasets ---")
    for i in range(m_datasets):
        completed = imputed_list.complete_data(dataset=i)

        completed["osm_acreage"] = completed["osm_acreage"].clip(lower=0)
        completed["Baseline_Value_Per_Acre"] = (
            completed["Baseline_Value_Per_Acre"].clip(lower=0)
        )

        out = acreage_df.copy()
        out["osm_acreage"] = completed["osm_acreage"].values
        out["Baseline_Value_Per_Acre"] = completed["Baseline_Value_Per_Acre"].values

        fname = out_dir / f"Py_Imputed_Dataset_{i + 1}.csv"
        out.to_csv(fname, index=False)
        del completed, out; gc.collect()
        print(f"    [OK] {fname}")

    # 5. Verification report (Dataset 1) -------------------------------------
    ds1 = pd.read_csv(out_dir / "Py_Imputed_Dataset_1.csv")

    print("\n=== IMPUTATION VERIFICATION (Dataset 1) ===")
    print(f"  Method: miceforest v{mf.__version__} (LightGBM Random Forest)")
    print(f"  Datasets generated:     {m_datasets}")
    print(f"  Iterations per dataset: 10")

    for col in IMPUTE_COLS:
        s = ds1[col]
        print(f"\n  {col}:")
        print(f"    Missing:  {s.isna().sum()}")
        print(f"    Min:      {s.min():>14,.2f}")
        print(f"    Median:   {s.median():>14,.2f}")
        print(f"    Mean:     {s.mean():>14,.2f}")
        print(f"    Max:      {s.max():>14,.2f}")
        print(f"    Negative: {(s < 0).sum()}")


def run_pooling(in_dir, out_csv, m_datasets=100):
    aggregates  = []
    within_vars = []

    print("--- 1  Loading imputed datasets and computing aggregates ---\n")

    for i in range(1, m_datasets + 1):
        path = in_dir / f"Py_Imputed_Dataset_{i}.csv"
        if not path.exists():
            raise FileNotFoundError(f"File not found: {path}")

        df = pd.read_csv(path)
        df["Total_Opportunity_Cost"] = df["osm_acreage"] * df["Baseline_Value_Per_Acre"]

        q_i   = df["Total_Opportunity_Cost"].sum()
        var_i = df["Total_Opportunity_Cost"].var(ddof=1)

        aggregates.append(q_i)
        within_vars.append(var_i)

        del df; gc.collect()
        print(f"  Dataset {i}:  ${q_i / 1e9:>10.3f} B")

    print("\n--- 2  Applying Rubin's Rules ---")

    aggregates  = np.array(aggregates)
    within_vars = np.array(within_vars)

    # [METHODOLOGY] Rubin's Rules pooling - q_bar is the pooled national estimate;
    #               v_t combines within- and between-imputation variance (Rubin 1987)
    q_bar = aggregates.mean()
    v_w   = within_vars.mean()
    v_b   = aggregates.var(ddof=1)
    v_t   = v_w + v_b + v_b / m_datasets
    se    = np.sqrt(v_t)
    ci95_lo = q_bar - 1.960 * se
    ci95_hi = q_bar + 1.960 * se
    ci99_lo = q_bar - 2.576 * se
    ci99_hi = q_bar + 2.576 * se

    print("\n=== RUBIN'S RULES RESULTS ===")
    print(f"  Pooled Aggregate National Value:  ${q_bar / 1e9:>10.3f} B")
    print(f"  Within-Imputation Variance (v_w): {v_w:.4e}")
    print(f"  Between-Imputation Variance (v_b):{v_b:.4e}")
    print(f"  Total Variance (v_t):             {v_t:.4e}")
    print(f"  Standard Error:                   ${se / 1e9:>10.3f} B")
    print(
        f"  99% Confidence Interval:          "
        f"${ci99_lo / 1e9:>10.3f} B - ${ci99_hi / 1e9:>10.3f} B"
    )
    print(
        f"  95% Confidence Interval:          "
        f"${ci95_lo / 1e9:>10.3f} B - ${ci95_hi / 1e9:>10.3f} B"
    )

    metrics = [
        ("Pooled Aggregate National Value ($)",  f"{q_bar:.2f}"),
        ("Pooled Aggregate National Value ($B)", f"{q_bar / 1e9:.3f}"),
        ("Within-Imputation Variance (v_w)",     f"{v_w:.4e}"),
        ("Between-Imputation Variance (v_b)",    f"{v_b:.4e}"),
        ("Total Variance (v_t)",                 f"{v_t:.4e}"),
        ("Standard Error ($)",                   f"{se:.2f}"),
        ("99% CI Lower ($B)",                    f"{ci99_lo / 1e9:.3f}"),
        ("99% CI Upper ($B)",                    f"{ci99_hi / 1e9:.3f}"),
        ("95% CI Lower ($B)",                    f"{ci95_lo / 1e9:.3f}"),
        ("95% CI Upper ($B)",                    f"{ci95_hi / 1e9:.3f}"),
    ] + [
        (f"Dataset {i} Aggregate ($B)", f"{aggregates[i - 1] / 1e9:.3f}")
        for i in range(1, m_datasets + 1)
    ]

    pooled_df = pd.DataFrame(metrics, columns=["Metric", "Value"])
    pooled_df.to_csv(out_csv, index=False)
    print(f"\n  [OK] Saved -> {out_csv.name}")

    return q_bar, se, ci95_lo, ci95_hi, ci99_lo, ci99_hi


def pool_acreage(x: np.ndarray) -> dict:
    q_bar = x.mean()
    v_b   = x.var(ddof=1)
    se    = np.sqrt(v_b + v_b / len(x))
    return {
        "mean":  q_bar,
        "sd_b":  np.sqrt(v_b),
        "ci95_lo": q_bar - 1.960 * se,
        "ci95_hi": q_bar + 1.960 * se,
        "ci99_lo": q_bar - 2.576 * se,
        "ci99_hi": q_bar + 2.576 * se,
    }


def run_acreage_summary(in_dir, out_csv, m_datasets=100):
    print("Computing total U.S. golf course footprint (pooled across imputations)\n")

    print("--- 1  Loading imputed datasets and computing acreage totals ---\n")

    national_totals = np.zeros(m_datasets)
    by_type_frames  = []

    for i in range(1, m_datasets + 1):
        path = in_dir / f"Py_Imputed_Dataset_{i}.csv"
        if not path.exists():
            raise FileNotFoundError(f"File not found: {path}")

        df = pd.read_csv(path)
        national_totals[i - 1] = df["osm_acreage"].sum()

        type_sums = (
            df.groupby("county_type", as_index=False)["osm_acreage"]
            .sum()
            .rename(columns={"osm_acreage": "acreage"})
        )
        type_sums["imputation"] = i
        by_type_frames.append(type_sums)

        del df; gc.collect()
        urban = type_sums.loc[type_sums["county_type"] == "Urban", "acreage"].iat[0]
        rural = type_sums.loc[type_sums["county_type"] == "Rural", "acreage"].iat[0]
        print(
            f"  Dataset {i}:  {round(national_totals[i - 1]):>12,} acres"
            f"  ({round(urban):,} Urban / {round(rural):,} Rural)"
        )

    print("\n--- 2  Pooling across imputations ---\n")

    # [METHODOLOGY] Rubin's Rules (acreage) - between-imputation variance only;
    #               within-variance is zero for a spatially fixed attribute
    nat_pool = pool_acreage(national_totals)

    all_by_type = pd.concat(by_type_frames, ignore_index=True)
    type_pool = (
        all_by_type.groupby("county_type")["acreage"]
        .apply(lambda x: pd.Series(pool_acreage(x.to_numpy())))
        .unstack()
        .reset_index()
        .rename(columns={"mean": "pooled_acres"})
        .sort_values("pooled_acres", ascending=False)
        .reset_index(drop=True)
    )

    print("=== NATIONAL ACREAGE RESULTS ===")
    print(f"  Total U.S. Golf Acreage:  {round(nat_pool['mean']):,} acres")
    print(f"  Between-Imputation SD:    {nat_pool['sd_b']:,.2f}")
    print(
        f"  99% CI:                   "
        f"{round(nat_pool['ci99_lo']):,} - {round(nat_pool['ci99_hi']):,} acres"
    )
    print(
        f"  95% CI:                   "
        f"{round(nat_pool['ci95_lo']):,} - {round(nat_pool['ci95_hi']):,} acres"
    )
    for _, row in type_pool.iterrows():
        print(f"  {row['county_type']:<20} {round(row['pooled_acres']):,} acres")

    national_row = pd.DataFrame([{
        "Category":          "National Total",
        "County_Type":       "All",
        "Pooled_Acres":      round(nat_pool["mean"], 2),
        "SD_Between":        round(nat_pool["sd_b"], 4),
        "CI_95_Lower_Acres": round(nat_pool["ci95_lo"], 2),
        "CI_95_Upper_Acres": round(nat_pool["ci95_hi"], 2),
        "CI_99_Lower_Acres": round(nat_pool["ci99_lo"], 2),
        "CI_99_Upper_Acres": round(nat_pool["ci99_hi"], 2),
    }])
    type_rows = pd.DataFrame({
        "Category":          "By County Type",
        "County_Type":       type_pool["county_type"],
        "Pooled_Acres":      type_pool["pooled_acres"].round(2),
        "SD_Between":        type_pool["sd_b"].round(4),
        "CI_95_Lower_Acres": type_pool["ci95_lo"].round(2),
        "CI_95_Upper_Acres": type_pool["ci95_hi"].round(2),
        "CI_99_Lower_Acres": type_pool["ci99_lo"].round(2),
        "CI_99_Upper_Acres": type_pool["ci99_hi"].round(2),
    })
    summary_df = pd.concat([national_row, type_rows], ignore_index=True)
    summary_df.to_csv(out_csv, index=False)
    print(f"  [OK] Saved -> {out_csv.name}")


# === 4. EXECUTION ===

if __name__ == "__main__":
    run_imputation(INPUT_CSV, OUT_DIR, M)
    print("\n=== STEP 2: RUBIN'S RULES POOLING ===")
    run_pooling(OUT_DIR, OUT_RUBINS_CSV, M)
    print("\n=== STEP 3: NATIONAL ACREAGE SUMMARY ===")
    run_acreage_summary(OUT_DIR, OUT_ACREAGE_CSV, M)

File: 2 - Work/Phase 3 Economic Merge and MICE Imputation/Phase_3.jl
# Purpose: Complete Phase 3 pipeline - MICE imputation (m=5) then Rubin's
#          Rules pooling to produce a national land-value estimate with 95%/99% CI.
# Inputs:  Phase 2 Spatial Polygons and True Acreage/Data/Julia/
#            Jl_Phase2_Acreage_Matched.csv
# Outputs: Data/Julia/Jl_Imputed_Dataset_{1..100}.csv
#          Data/Julia/Jl_Rubins_Rules_Summary.csv
#          Data/Julia/Jl_National_Acreage_Summary.csv
#
# Rubin's Rules formula summary (m = 100 imputations):
#   q_bar = mean(Q_i)           -- pooled point estimate
#   v_w   = mean(Var_i)         -- within-imputation variance
#   v_b   = var(Q_i, ddof=1)    -- between-imputation variance
#   v_t   = v_w + v_b + v_b/m  -- total variance
#   se    = sqrt(v_t)
#   99%CI = q_bar +/- 2.576 * se
#   95%CI = q_bar +/- 1.960 * se


# === 1. LIBRARIES ===

using CategoricalArrays, CSV, DataFrames, Mice, Printf, Random, Statistics


# === 2. GLOBALS & PATHS ===

const SCRIPT_DIR = @__DIR__
const INPUT_CSV  = joinpath(
    SCRIPT_DIR, "..",
    "Phase 2 Spatial Polygons and True Acreage",
    "Data", "Julia",
    "Jl_Phase2_Acreage_Matched.csv"
)
const OUT_DIR = joinpath(SCRIPT_DIR, "Data", "Julia")
const OUT_CSV         = joinpath(OUT_DIR, "Jl_Rubins_Rules_Summary.csv")
const OUT_ACREAGE_CSV = joinpath(OUT_DIR, "Jl_National_Acreage_Summary.csv")
const M             = 100
const IMPUTE_COLS   = [:osm_acreage, :Baseline_Value_Per_Acre]
const PREDICTOR_COLS = [:Holes, :Course_Type, :county_type, :Longitude, :Latitude]


# === 3. FUNCTIONS ===

function run_imputation(input_csv::String, out_dir::String; m_datasets::Int = 5)
    # 1. Load ----------------------------------------------------------------
    println("--- 1  Loading Phase 2 acreage-matched dataset ---")

    isfile(input_csv) || error("Input file not found: $input_csv")
    mkpath(out_dir)

    acreage_df = CSV.read(input_csv, DataFrame)
    println("    Rows: $(size(acreage_df, 1))")
    println("    Missing osm_acreage:            $(sum(ismissing, acreage_df.osm_acreage))")
    println("    Missing Baseline_Value_Per_Acre: $(sum(ismissing, acreage_df.Baseline_Value_Per_Acre))")

    # 2. Prepare -------------------------------------------------------------
    println("--- 2  Preparing imputation frame ---")

    acreage_df.Holes = [
        ismissing(x) || string(x) == "" ? missing :
        parse(Float64, string(x))
        for x in acreage_df.Holes
    ]
    acreage_df.Course_Type = categorical(acreage_df.Ownership_Type)
    acreage_df.county_type = categorical(acreage_df.county_type)

    model_cols = vcat(PREDICTOR_COLS, IMPUTE_COLS)
    imp_df     = acreage_df[!, model_cols]

    # 3. Run MICE ------------------------------------------------------------
    println("--- 3  Running MICE imputation (m=$m_datasets) ---")
    Random.seed!(42)  # [METHODOLOGY] reproducibility seed for stochastic MICE imputation
    # [METHODOLOGY] Mice.jl MICE - m=100 multiply-imputed datasets, iter=10;
    #               see Van Buuren (2018) for methodology
    imputed_list = mice(imp_df, m = m_datasets, iter = 10)

    # 4. Save each imputed dataset -------------------------------------------
    println("\n--- 4  Saving $m_datasets imputed datasets ---")

    for i in 1:m_datasets
        completed = complete(imputed_list, i)

        completed.osm_acreage            = clamp.(completed.osm_acreage, 0, Inf)
        completed.Baseline_Value_Per_Acre = clamp.(
            completed.Baseline_Value_Per_Acre, 0, Inf
        )

        out = copy(acreage_df)
        out.osm_acreage            = completed.osm_acreage
        out.Baseline_Value_Per_Acre = completed.Baseline_Value_Per_Acre

        # Mice.jl's complete() returns drawn values for ALL rows, not just missing
        # ones. Restore original observed (non-missing) values so MICE never
        # overwrites anchored inputs like Hawaii Kai's BVPA = $4,952,600.
        for col in IMPUTE_COLS
            orig = acreage_df[!, col]
            obs  = .!ismissing.(orig)
            out[obs, col] = orig[obs]
        end

        fname = joinpath(out_dir, "Jl_Imputed_Dataset_$i.csv")
        CSV.write(fname, out; header = true, quote_empty_string = false)
        println("    [OK] $fname")
    end

    # 5. Verification report (Dataset 1) -------------------------------------
    ds1 = CSV.read(joinpath(out_dir, "Jl_Imputed_Dataset_1.csv"), DataFrame)

    println("\n=== IMPUTATION VERIFICATION (Dataset 1) ===")
    println("  Method: Mice.jl")
    println("  Datasets generated:     $m_datasets")
    println("  Iterations per dataset: 10")

    for col in IMPUTE_COLS
        s = ds1[!, col]
        println("\n  $col:")
        println("    Missing:  $(sum(ismissing, s))")
        @printf "    Min:      %14.2f\n" minimum(s)
        @printf "    Median:   %14.2f\n" median(s)
        @printf "    Mean:     %14.2f\n" mean(s)
        @printf "    Max:      %14.2f\n" maximum(s)
        println("    Negative: $(sum(<(0), s))")
    end
end

function run_pooling(in_dir::String, out_csv::String; m_datasets::Int = 5)
    aggregates  = Float64[]
    within_vars = Float64[]

    println("--- 1  Loading imputed datasets and computing aggregates ---\n")

    for i in 1:m_datasets
        fpath = joinpath(in_dir, "Jl_Imputed_Dataset_$i.csv")
        isfile(fpath) || error("Imputed dataset not found: $fpath")

        df = CSV.read(fpath, DataFrame)
        df.Total_Opportunity_Cost = df.osm_acreage .* df.Baseline_Value_Per_Acre

        q_i   = sum(df.Total_Opportunity_Cost)
        var_i = var(df.Total_Opportunity_Cost)

        push!(aggregates,  q_i)
        push!(within_vars, var_i)

        df = nothing; GC.gc()
        @printf("  Dataset %d:  \$%10.3f B\n", i, q_i / 1e9)
    end

    # [METHODOLOGY] Rubin's Rules pooling - q_bar is the pooled national estimate;
    #               v_t combines within- and between-imputation variance (Rubin 1987)
    println("\n--- 2  Applying Rubin's Rules ---")

    q_bar = mean(aggregates)
    v_w   = mean(within_vars)
    v_b   = var(aggregates; corrected = true)
    v_t   = v_w + v_b + v_b / m_datasets
    se    = sqrt(v_t)
    ci95_lo = q_bar - 1.960 * se
    ci95_hi = q_bar + 1.960 * se
    ci99_lo = q_bar - 2.576 * se
    ci99_hi = q_bar + 2.576 * se

    println("\n=== RUBIN'S RULES RESULTS ===")
    @printf("  Pooled Aggregate National Value:  \$%10.3f B\n", q_bar / 1e9)
    @printf("  Within-Imputation Variance (v_w): %.4e\n", v_w)
    @printf("  Between-Imputation Variance (v_b):%.4e\n", v_b)
    @printf("  Total Variance (v_t):             %.4e\n", v_t)
    @printf("  Standard Error:                   \$%10.3f B\n", se / 1e9)
    @printf(
        "  99%% Confidence Interval:          \$%10.3f B - \$%10.3f B\n",
        ci99_lo / 1e9, ci99_hi / 1e9
    )
    @printf(
        "  95%% Confidence Interval:          \$%10.3f B - \$%10.3f B\n",
        ci95_lo / 1e9, ci95_hi / 1e9
    )

    pooled_df = DataFrame(
        Metric = vcat(
            [
                "Pooled Aggregate National Value (\$)",
                "Pooled Aggregate National Value (\$B)",
                "Within-Imputation Variance (v_w)",
                "Between-Imputation Variance (v_b)",
                "Total Variance (v_t)",
                "Standard Error (\$)",
                "99% CI Lower (\$B)",
                "99% CI Upper (\$B)",
                "95% CI Lower (\$B)",
                "95% CI Upper (\$B)",
            ],
            ["Dataset $i Aggregate (\$B)" for i in 1:m_datasets]
        ),
        Value = vcat(
            [
                @sprintf("%.2f",  q_bar),
                @sprintf("%.3f",  q_bar / 1e9),
                @sprintf("%.4e",  v_w),
                @sprintf("%.4e",  v_b),
                @sprintf("%.4e",  v_t),
                @sprintf("%.2f",  se),
                @sprintf("%.3f",  ci99_lo / 1e9),
                @sprintf("%.3f",  ci99_hi / 1e9),
                @sprintf("%.3f",  ci95_lo / 1e9),
                @sprintf("%.3f",  ci95_hi / 1e9),
            ],
            [@sprintf("%.3f", aggregates[i] / 1e9) for i in 1:m_datasets]
        )
    )

    CSV.write(out_csv, pooled_df; header = true)
    println("\n  [OK] Saved -> $out_csv")

    return q_bar, se, ci95_lo, ci95_hi, ci99_lo, ci99_hi
end


function pool_acreage(x::AbstractVector{<:Real})
    q_bar = mean(x)
    v_b   = var(x; corrected = true)
    se    = sqrt(v_b + v_b / length(x))
    return (
        mean    = q_bar,
        sd_b    = sqrt(v_b),
        ci95_lo = q_bar - 1.960 * se,
        ci95_hi = q_bar + 1.960 * se,
        ci99_lo = q_bar - 2.576 * se,
        ci99_hi = q_bar + 2.576 * se,
    )
end

function run_acreage_summary(in_dir::String, out_csv::String; m_datasets::Int = 5)
    fmt(x) = replace(@sprintf("%d", round(Int, x)), r"(?<=\d)(?=(\d{3})+$)" => ",")

    println("--- 1  Loading imputed datasets and computing acreage totals ---\n")

    national_totals = zeros(Float64, m_datasets)
    by_type_list    = Vector{DataFrame}(undef, m_datasets)

    for i in 1:m_datasets
        path = joinpath(in_dir, "Jl_Imputed_Dataset_$i.csv")
        isfile(path) || error("File not found: $path")

        df = CSV.read(path, DataFrame)

        national_totals[i] = sum(skipmissing(df.osm_acreage))

        type_sums = combine(
            groupby(df, :county_type),
            :osm_acreage => (x -> sum(skipmissing(x))) => :acreage,
        )
        type_sums.imputation = fill(i, nrow(type_sums))
        by_type_list[i] = type_sums

        df = nothing; GC.gc()
        urban_acres = only(filter(r -> isequal(r.county_type, "Urban"), type_sums).acreage)
        rural_acres = only(filter(r -> isequal(r.county_type, "Rural"), type_sums).acreage)
        @printf("  Dataset %d:  %s acres  (%s Urban / %s Rural)\n",
            i, fmt(national_totals[i]), fmt(urban_acres), fmt(rural_acres))
    end

    println("\n--- 2  Pooling across imputations ---\n")

    # [METHODOLOGY] Rubin's Rules (acreage) - between-imputation variance only;
    #               within-variance is zero for a spatially fixed attribute
    nat_pool    = pool_acreage(national_totals)
    all_by_type = vcat(by_type_list...)
    type_groups = groupby(all_by_type, :county_type)
    type_pool   = combine(type_groups) do grp
        p = pool_acreage(grp.acreage)
        DataFrame(
            pooled_acres = p.mean,
            sd_b         = p.sd_b,
            ci95_lo      = p.ci95_lo,
            ci95_hi      = p.ci95_hi,
            ci99_lo      = p.ci99_lo,
            ci99_hi      = p.ci99_hi,
        )
    end
    sort!(type_pool, :pooled_acres, rev = true)

    println("=== NATIONAL ACREAGE RESULTS ===")
    @printf("  Total U.S. Golf Acreage:  %s acres\n",          fmt(nat_pool.mean))
    @printf("  Between-Imputation SD:    %.2f\n",               nat_pool.sd_b)
    @printf("  99%% CI:                   %s - %s acres\n",     fmt(nat_pool.ci99_lo), fmt(nat_pool.ci99_hi))
    @printf("  95%% CI:                   %s - %s acres\n",     fmt(nat_pool.ci95_lo), fmt(nat_pool.ci95_hi))
    for row in eachrow(type_pool)
        @printf("  %-20s %s acres\n", row.county_type, fmt(row.pooled_acres))
    end

    national_row = DataFrame(
        Category          = "National Total",
        County_Type       = "All",
        Pooled_Acres      = round(nat_pool.mean, digits = 2),
        SD_Between        = round(nat_pool.sd_b,  digits = 4),
        CI_95_Lower_Acres = round(nat_pool.ci95_lo, digits = 2),
        CI_95_Upper_Acres = round(nat_pool.ci95_hi, digits = 2),
        CI_99_Lower_Acres = round(nat_pool.ci99_lo, digits = 2),
        CI_99_Upper_Acres = round(nat_pool.ci99_hi, digits = 2),
    )
    type_rows = DataFrame(
        Category          = fill("By County Type", nrow(type_pool)),
        County_Type       = type_pool.county_type,
        Pooled_Acres      = round.(type_pool.pooled_acres, digits = 2),
        SD_Between        = round.(type_pool.sd_b,         digits = 4),
        CI_95_Lower_Acres = round.(type_pool.ci95_lo,      digits = 2),
        CI_95_Upper_Acres = round.(type_pool.ci95_hi,      digits = 2),
        CI_99_Lower_Acres = round.(type_pool.ci99_lo,      digits = 2),
        CI_99_Upper_Acres = round.(type_pool.ci99_hi,      digits = 2),
    )
    summary_df = vcat(national_row, type_rows)

    CSV.write(out_csv, summary_df)
    println("\n  [OK] Saved -> $out_csv")
end


# === 4. EXECUTION ===

function main()
    println("PHASE 3 ANALYSIS PIPELINE (Julia)")
    println("Output directory: $OUT_DIR\n")

    run_imputation(INPUT_CSV, OUT_DIR; m_datasets = M)

    println("\n=== STEP 2: RUBIN'S RULES POOLING ===")
    run_pooling(OUT_DIR, OUT_CSV; m_datasets = M)

    println("\n=== STEP 3: NATIONAL ACREAGE SUMMARY ===")
    run_acreage_summary(OUT_DIR, OUT_ACREAGE_CSV; m_datasets = M)

    println("\n=== PHASE 3 ANALYSIS COMPLETE ===")
    println("Output files saved to: $OUT_DIR")
    for i in 1:M
        println("  - Jl_Imputed_Dataset_$i.csv")
    end
    println("  - Jl_Rubins_Rules_Summary.csv")
    println("  - Jl_National_Acreage_Summary.csv")
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
